### PR TITLE
IGNITE-16593 Schema should keep colocation information

### PR DESCRIPTION
**Pull request by tledkov-gridgain**

### CODE REVIEW

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/definition/builder/PrimaryKeyDefinitionBuilderImpl.java [hunk 1]
```diff
@@ -98,12 +98,12 @@ public PrimaryKeyDefinition build() {
...
 
-        Set<String> affCols;
+        List<String> affCols;
 
         if (CollectionUtils.nullOrEmpty(affinityColumns)) {
-            affCols = cols;
+            affCols = List.copyOf(columns);
         } else {
-            affCols = Set.copyOf(affinityColumns);
+            affCols = List.copyOf(affinityColumns);
```
Comment (ygerzhedovich) [inline modules/schema/src/main/java/org/apache/ignite/internal/schema/definition/builder/PrimaryKeyDefinitionBuilderImpl.java:106]: Earlier we took only unique columns. Should we check that we don't use the same coumns twise?

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/configuration/SchemaConfigurationConverter.java [hunk 2]
```diff
@@ -230,7 +230,7 @@ public static IndexDefinition convert(TableIndexView idxView) {
      * @return TableIn.
      */
     public static PrimaryKeyDefinition convert(PrimaryKeyView primaryKey) {
-        return new PrimaryKeyDefinitionImpl(Set.of(primaryKey.columns()), Set.of(primaryKey.affinityColumns()));
+        return new PrimaryKeyDefinitionImpl(Set.of(primaryKey.columns()), List.of(primaryKey.colocationColumns()));
```
Comment (AMashenkov) [inline modules/schema/src/main/java/org/apache/ignite/internal/schema/configuration/SchemaConfigurationConverter.java:233]: Colocation columns is not check for uniquiness here.

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/definition/builder/PrimaryKeyDefinitionBuilderImpl.java (hunk 1 above)
Comment (AMashenkov) [reply modules/schema/src/main/java/org/apache/ignite/internal/schema/definition/builder/PrimaryKeyDefinitionBuilderImpl.java]: @ygerzhedovich, affinityColumns are checked few lines below.
But affinityColumns are not checked in SchemaConfigurationConverter.

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/configuration/SchemaConfigurationConverter.java (hunk 2 above)
Comment (tledkov-gridgain) [reply modules/schema/src/main/java/org/apache/ignite/internal/schema/configuration/SchemaConfigurationConverter.java:233]: Fixed. Check moved to constructor of fhe `PrimaryKeyDefinitionImpl`

File: modules/api/src/main/java/org/apache/ignite/schema/definition/builder/PrimaryKeyDefinitionBuilder.java [hunk 3]
```diff
@@ -30,15 +30,15 @@
      * @param cols Affinity columns. Must be a valid subset of key columns.
      * @return Primary index builder.
      */
-    PrimaryKeyDefinitionBuilder withAffinityColumns(String... cols);
+    PrimaryKeyDefinitionBuilder withColocationColumns(String... cols);
 
     /**
      * Sets affinity columns.
      *
      * @param cols Affinity columns. Must be a valid subset of key columns.
```
Comment (ygerzhedovich) [inline modules/api/src/main/java/org/apache/ignite/schema/definition/builder/PrimaryKeyDefinitionBuilder.java:38]: Affinity -> Colocation

File: modules/api/src/main/java/org/apache/ignite/schema/definition/builder/PrimaryKeyDefinitionBuilder.java [hunk 4]
```diff
@@ -30,15 +30,15 @@
...
```
Comment (ygerzhedovich) [inline modules/api/src/main/java/org/apache/ignite/schema/definition/builder/PrimaryKeyDefinitionBuilder.java:30]: Affinity -> Colocation

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/definition/TableDefinitionImpl.java [hunk 5]
```diff
@@ -51,7 +51,7 @@
     private final Set<String> keyCols;
 
     /** Cached key affinity columns. */
```
Comment (ygerzhedovich) [inline modules/schema/src/main/java/org/apache/ignite/internal/schema/definition/TableDefinitionImpl.java:53]: Affinity -> Colocation

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/definition/builder/PrimaryKeyDefinitionBuilderImpl.java [hunk 6]
```diff
@@ -63,8 +63,8 @@ public PrimaryKeyDefinitionBuilderImpl withColumns(List<String> columns) {
 
     /** {@inheritDoc} */
     @Override
-    public PrimaryKeyDefinitionBuilderImpl withAffinityColumns(String... affinityColumns) {
-        this.affinityColumns = affinityColumns == null
+    public PrimaryKeyDefinitionBuilderImpl withColocationColumns(String... affinityColumns) {
```
Comment (ygerzhedovich) [inline modules/schema/src/main/java/org/apache/ignite/internal/schema/definition/builder/PrimaryKeyDefinitionBuilderImpl.java:66]: affinityCollumns -> colocationColumns

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/definition/builder/PrimaryKeyDefinitionBuilderImpl.java [hunk 7]
```diff
@@ -73,8 +73,8 @@ public PrimaryKeyDefinitionBuilderImpl withAffinityColumns(String... affinityCol
 
     /** {@inheritDoc} */
     @Override
-    public PrimaryKeyDefinitionBuilderImpl withAffinityColumns(List<String> affinityColumns) {
-        this.affinityColumns = affinityColumns == null
+    public PrimaryKeyDefinitionBuilderImpl withColocationColumns(List<String> affinityColumns) {
```
Comment (ygerzhedovich) [inline modules/schema/src/main/java/org/apache/ignite/internal/schema/definition/builder/PrimaryKeyDefinitionBuilderImpl.java:76]: affinity.... -> colocation....

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/schema/SchemaSerializerImpl.java [hunk 8]
```diff
@@ -89,7 +89,7 @@ public void writeTo(SchemaDescriptor desc, ByteBuffer byteBuf) {
         appendColumns(desc.keyColumns(), byteBuf);
         appendColumns(desc.valueColumns(), byteBuf);
 
-        Column[] affinityCols = desc.affinityColumns();
+        Column[] affinityCols = desc.colocationColumns();
```
Comment (ygerzhedovich) [inline modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/schema/SchemaSerializerImpl.java:92]: affinity -> colocation

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/schema/SchemaSerializerImpl.java [hunk 9]
```diff
@@ -133,7 +133,7 @@ public int size(SchemaDescriptor desc) {
                 + getColumnsSize(desc.keyColumns())
                 + getColumnsSize(desc.valueColumns())
                 + ARRAY_HEADER_LENGTH          //Affinity columns length
```
Comment (ygerzhedovich) [inline modules/schema/src/main/java/org/apache/ignite/internal/schema/marshaller/schema/SchemaSerializerImpl.java:135]: Affinity -> Colocation

File: modules/schema/src/test/java/org/apache/ignite/internal/schema/SchemaConfigurationTest.java [hunk 10]
```diff
@@ -51,7 +53,7 @@ public void testInitialSchema() {
                 .withPrimaryKey(
                         SchemaBuilders.primaryKey()  // Declare index column in order.
                                 .withColumns("id", "affId", "name")
-                                .withAffinityColumns(
+                                .withColocationColumns(
                                         "affId") // Optional affinity declaration. If not set, all columns will be affinity cols.
```
Comment (ygerzhedovich) [inline modules/schema/src/test/java/org/apache/ignite/internal/schema/SchemaConfigurationTest.java:57]: Affinity -> Colocation

File: modules/schema/src/test/java/org/apache/ignite/internal/schema/builder/PrimaryKeyDefinitionDefinitionBuilderTest.java [hunk 11]
```diff
@@ -37,18 +37,18 @@
     public void testPrimaryKeyWithAffinityColumns() {
```
Comment (ygerzhedovich) [inline modules/schema/src/test/java/org/apache/ignite/internal/schema/builder/PrimaryKeyDefinitionDefinitionBuilderTest.java:37]: Affinity -> Colocation

File: modules/schema/src/test/java/org/apache/ignite/internal/schema/builder/PrimaryKeyDefinitionDefinitionBuilderTest.java [hunk 12]
```diff
@@ -61,18 +61,18 @@ public void testPrimaryKey() {
...
+        assertEquals(3, idx.colocationColumns().size());
 
         assertTrue(idx.columns().containsAll(List.of("A", "B", "C")));
-        assertTrue(idx.affinityColumns().containsAll(List.of("A", "B", "C")));
+        assertTrue(idx.colocationColumns().containsAll(List.of("A", "B", "C")));
     }
 
     /** Test primary key parameters. */
     @Test
     public void testPrimaryKeyWrongAffinityColumn() {
```
Comment (ygerzhedovich) [inline modules/schema/src/test/java/org/apache/ignite/internal/schema/builder/PrimaryKeyDefinitionDefinitionBuilderTest.java:72]: Affinity -> Colocation

**Review by AMashenkov — APPROVED** [review]

**Pull request merged**